pkg/cloudprovider/providers/vsphere/nodeuuidcache.go

```diff
@@ -0,0 +1,184 @@
+	}
+
+	_, err := cache.loadFromConfigMap()
+	if err != nil {
+		return "", fmt.Errorf("error loading uuid in configmap %s with %v", cache.configMapName, err)
+	}
+	if vmUUID, ok := cache.nodeUUIDMap[nodeName]; ok {
+		return vmUUID.uuid, nil
+	}
+	return "", fmt.Errorf("error finding uuid for %s", nodeName)
+}
+
+func (cache *nodeUUIDCache) loadFromConfigMap() (*v1.ConfigMap, error) {
+	if cache.kubeClient == nil {
+		return nil, fmt.Errorf("kubeClient is not initialized yet")
+	}
+	var configMap *v1.ConfigMap
+	var err error
+	configMap, err = cache.kubeClient.CoreV1().ConfigMaps(cache.namespace).Get(cache.configMapName, metav1.GetOptions{})
+
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, fmt.Errorf("error loading vsphere-uuid configmap %s with %v", cache.configMapName, err)
+		}
+		initializeErr := cache.initializeConfigMap()
+		if initializeErr != nil {
+			return nil, initializeErr
+		}
+		configMap, err = cache.kubeClient.CoreV1().ConfigMaps(cache.namespace).Get(cache.configMapName, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("error loading vsphere-uuid configmap %s with %v", cache.configMapName, err)
+		}
+
+	}
+	configMapData := configMap.Data
+	for k, v := range configMapData {
+		if _, ok := cache.nodeUUIDMap[k]; !ok {
+			// add all new entries loaded from configmap as not deleted
+			cache.nodeUUIDMap[k] = &nodeUUID{uuid: v}
+		}
+	}
+	return configMap, nil
+}
+
+func (cache *nodeUUIDCache) initializeConfigMap() error {
+	if cache.kubeClient == nil {
+		return fmt.Errorf("kubeClient is not initialized yet")
+	}
+	_, err := cache.kubeClient.CoreV1().ConfigMaps(cache.namespace).Create(&v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: cache.configMapName,
+		},
+		Data: map[string]string{},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to initialize vsphere-uuid configmap %s with %v", cache.configMapName, err)
+	}
+	return nil
+}
+
+func (cache *nodeUUIDCache) updateConfigMap(configMap *v1.ConfigMap) error {
+	_, err := cache.kubeClient.CoreV1().ConfigMaps(cache.namespace).Update(configMap)
+	if err != nil {
+		glog.Errorf("error while updating vm uuid configmap. err: %+v", err)
+		return err
+	}
+	return nil
+}
+
+func (cache *nodeUUIDCache) getUUIData() map[string]string {
+	// if we have too many entries we should prune not deleted node entries
+	if len(cache.nodeUUIDMap) > maxDataEntries {
+		cache.pruneCache()
+	}
+	result := make(map[string]string, len(cache.nodeUUIDMap))
+	for k, v := range cache.nodeUUIDMap {
+		// Even after pruning if cache still has more than maxDataEntries
+		// then we just pick first N values. Not ideal - we are limited
+		// by size limit of configmaps.
+		if len(result) > maxDataEntries {
+			return result
+		}
+		result[k] = v.uuid
+	}
+	return result
+}
+
+func (cache *nodeUUIDCache) pruneCache() {
+	for k, v := range cache.nodeUUIDMap {
+		// if node is not deleted then we can still fetch node's UUID from node object itself
+		if !v.deleted {
```

Comment:
When a node is deleted NodeDeleted method is called where the latest node object is passed. This object has field VolumeAttached which can be used to know if volumes are still attached to the node. With this information, nodes with no volume attached can be deleted first.

Reply:
Discussed this offline, I will file a separate issue for handling more than 3600 nodes in more robust way.